test/controllers/dashboard_controller_test.rb

```diff
@@ -133,6 +137,63 @@ def dropdown_list_items(list)
     assert_select "nav a[href='#{batch_connect_sessions_path}']", 0
   end
 
+  #test whitelist mode 
+  test "should not create app menus if NavConfig.categories is empty and whitelist is enabled" do
+    SysRouter.stubs(:base_path).returns(Rails.root.join("test/fixtures/sys_with_gateway_apps"))
+    OodAppkit.stubs(:clusters).returns(OodCore::Clusters.load_file("test/fixtures/config/clusters.d"))
+    NavConfig.stubs(:categories_whitelist).returns(true)
+    NavConfig.stubs(:categories).returns([])
+    
+    get :index
+    assert_response :success
+    assert_select ".navbar-collapse > .nav li.dropdown[title]", 0
+  end
+
+  test "should exclude gateway apps if NavConfig.categories is set to default and whitelist is enabled" do
+    SysRouter.stubs(:base_path).returns(Rails.root.join("test/fixtures/sys_with_gateway_apps"))
+    OodAppkit.stubs(:clusters).returns(OodCore::Clusters.load_file("test/fixtures/config/clusters.d"))
+    NavConfig.stubs(:categories_whitelist).returns(true)
+    NavConfig.stubs(:categories).returns(["Files", "Jobs", "Clusters", "Interactive Apps"])
+
+    get :index
+    assert_response :success
+    assert_select ".navbar-collapse > .nav li.dropdown[title='Gateway Apps']", 0
+  end
+
+  test "should include gateway and System Installed apps in alphabetical order in end if NavConfig.categories is set to default listing and whitelist is disabled" do
+    SysRouter.stubs(:base_path).returns(Rails.root.join("test/fixtures/sys_with_gateway_apps"))
+    OodAppkit.stubs(:clusters).returns(OodCore::Clusters.load_file("test/fixtures/config/clusters.d"))
+    NavConfig.stubs(:categories_whitelist).returns(false)
+    NavConfig.stubs(:categories).returns(["Files", "Jobs", "Clusters", "Interactive Apps"])
+    
+    get :index
+    assert_response :success
+    assert_select ".navbar-collapse > .nav li.dropdown[title]", 6
+    assert_select  dropdown_link(1), text: "Files"
+    assert_select  dropdown_link(2), text: "Jobs"
+    assert_select  dropdown_link(3), text: "Clusters"
+    assert_select  dropdown_link(4), text: "Interactive Apps"
+    assert_select  dropdown_link(5), text: "Gateway Apps"
+    assert_select  dropdown_link(6), text: "System Installed Apps"
```

Comment:
Whenever you stub class methods I think you need to unstub them at the end or else their stubbing will remain in future tests.

Reply:
Maybe. Actually could you research that? It looks like I don't unstub in some other tests.

Reply:
I didn't find the answer from documentation about stubbing life cycle. But I created a new test method and it seems stubbed navConfig variables from the previous test methods are gone. I think the stubbed value in the test method won't carry over to future tests. However, should I still call unstub at end of the test methods just in case?